python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -0,0 +1,252 @@
+import itertools
+from typing import Callable, List, Optional, Tuple
+
+import ray
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data._internal.split import _split_at_indices
+from ray.data._internal.stats import StatsDict
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data.block import (
+    Block,
+    BlockAccessor,
+    BlockExecStats,
+    BlockMetadata,
+    BlockPartition,
+)
+
+
+class ZipOperator(PhysicalOperator):
+    """An operator that zips its inputs together.
+
+    NOTE: the implementation is bulk for now, which materializes all its inputs in
+    object store, before starting execution. Should re-implement it as a streaming
+    operator in the future.
+    """
+
+    def __init__(
+        self,
+        left_input_op: PhysicalOperator,
+        right_input_op: PhysicalOperator,
+    ):
+        """Create a ZipOperator.
+
+        Args:
+            left_input_ops: The input operator at left hand side.
+            right_input_op: The input operator at right hand side.
+        """
+        self._left_buffer: List[RefBundle] = []
+        self._right_buffer: List[RefBundle] = []
+        self._output_buffer: List[RefBundle] = []
+        self._stats: StatsDict = {}
+        super().__init__("Zip", [left_input_op, right_input_op])
+
+    def num_outputs_total(self) -> Optional[int]:
+        left_num_outputs = self.input_dependencies[0].num_outputs_total()
+        right_num_outputs = self.input_dependencies[1].num_outputs_total()
+        if left_num_outputs is not None and right_num_outputs is not None:
+            return max(left_num_outputs, right_num_outputs)
```

Comment:
Since we require the num rows to be the same, should this assert them equal and return just one of them (instead of max)?

Reply:
@jianoaix - this is total number of output bundles, right? Output bundles may not necessarilly be same, even though the number of rows are same. So feel it's not able to assert here. Let me know if my understanding is wrong.

Reply:
Oh, right, they are not the same. I guess the issue might be the  bock splitting in `_zip()` that make the `max` inaccurate.